<a href="https://colab.research.google.com/github/SK-0911/YBI-BDA-Internship/blob/main/YBI_Internship_Class(BDA).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('titanic.csv')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
df.describe(include='all')

PassengerId    Survived      Pclass                     Name   Sex  \
count    891.000000  891.000000  891.000000                      891   891   
unique          NaN         NaN         NaN                      891     2   
top             NaN         NaN         NaN  Braund, Mr. Owen Harris  male   
freq            NaN         NaN         NaN                        1   577   
mean     446.000000    0.383838    2.308642                      NaN   NaN   
std      257.353842    0.486592    0.836071                      NaN   NaN   
min        1.000000    0.000000    1.000000                      NaN   NaN   
25%      223.500000    0.000000    2.000000                      NaN   NaN   
50%      446.000000    0.000000    3.000000                      NaN   NaN   
75%      668.500000    1.000000    3.000000                      NaN   NaN   
max      891.000000    1.000000    3.000000                      NaN   NaN   

               Age       SibSp       Parch  Ticket        Fare    Cabin  \
count   714.000000  891.000000  891.000000     891  891.000000      204   
unique         NaN         NaN         NaN     681         NaN      147   
top            NaN         NaN         NaN  347082         NaN  B96 B98   
freq           NaN         NaN         NaN       7         NaN        4   
mean     29.699118    0.523008    0.381594     NaN   32.204208      NaN   
std      14.526497    1.102743    0.806057     NaN   49.693429      NaN   
min       0.420000    0.000000    0.000000     NaN    0.000000      NaN   
25%      20.125000    0.000000    0.000000     NaN    7.910400      NaN   
50%      28.000000    0.000000    0.000000     NaN   14.454200      NaN   
75%      38.000000    1.000000    0.000000     NaN   31.000000      NaN   
max      80.000000    8.000000    6.000000     NaN  512.329200      NaN   

       Embarked  
count       889  
unique        3  
top           S  
freq        644  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

In [ ]:
df.shape

(891, 12)

In [ ]:
df['Sex'].unique()

array(['male', 'female'], dtype=object)

In [ ]:
df['Sex'].nunique()

2

In [ ]:
df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

# 8 steps of Data analytics



## Step 1: Import Libraries



In [ ]:
import pandas as pd

## Step 2: Read Data

In [ ]:
salary = pd.read_csv('https://github.com/YBI-Foundation/Dataset/raw/main/Salary%20Data.csv')

In [ ]:
salary

Experience Years  Salary
0                1.1   39343
1                1.2   42774
2                1.3   46205
3                1.5   37731
4                2.0   43525
5                2.2   39891
6                2.5   48266
7                2.9   56642
8                3.0   60150
9                3.2   54445
10               3.2   64445
11               3.5   60000
12               3.7   57189
13               3.8   60200
14               3.9   63218
15               4.0   55794
16               4.0   56957
17               4.1   57081
18               4.3   59095
19               4.5   61111
20               4.7   64500
21               4.9   67938
22               5.1   66029
23               5.3   83088
24               5.5   82200
25               5.9   81363
26               6.0   93940
27               6.2   91000
28               6.5   90000
29               6.8   91738
30               7.1   98273
31               7.9  101302
32               8.2  113812
33               8.5  111620
34               8.7  109431
35               9.0  105582
36               9.5  116969
37               9.6  112635
38              10.3  122391
39              10.5  121872

In [ ]:
salary.head()

Experience Years  Salary
0               1.1   39343
1               1.2   42774
2               1.3   46205
3               1.5   37731
4               2.0   43525

In [ ]:
salary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Experience Years  40 non-null     float64
 1   Salary            40 non-null     int64  
dtypes: float64(1), int64(1)
memory usage: 768.0 bytes


## Step 3: Define target (y) and features(x)

In [ ]:
salary.columns

Index(['Experience Years', 'Salary'], dtype='object')

In [ ]:
y = salary['Salary']

y.shape

(40,)

In [ ]:
x = salary[['Experience Years']]

x.shape

(40, 1)

In [ ]:
salary.shape

(40, 2)

## Step 4: Split data into training and testing

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x,y, train_size=0.75, random_state=2)

x_train

Experience Years
38              10.3
29               6.8
24               5.5
12               3.7
16               4.0
1                1.2
33               8.5
23               5.3
25               5.9
19               4.5
32               8.2
10               3.2
4                2.0
6                2.5
3                1.5
34               8.7
5                2.2
28               6.5
20               4.7
26               6.0
39              10.5
21               4.9
35               9.0
31               7.9
7                2.9
11               3.5
18               4.3
22               5.1
8                3.0
15               4.0

In [ ]:
x_train.shape

(30, 1)

## Step 5: Select Model

In [ ]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

## Step 6: Train Model (fit model)

In [ ]:
model.fit(x_train, y_train)

LinearRegression()

In [ ]:
model.intercept_

24540.042250099512

In [ ]:
model.coef_

array([9687.61825014])

So, the equation of line is 

Salary = 24540 + (9688 * experience) + **error**

## Step 7: Predict

In [ ]:
y_pred = model.predict(x_test)
y_pred

array([ 84603.27540097,  55540.42065055,  62321.75342565,  35196.42232525,
        37133.94597528,  93322.1318261 ,  61352.99160063, 116572.41562644,
        64259.27707568, 117541.17745145])

## Step 8: Calculate Accuracy

In [ ]:
from sklearn.metrics import mean_absolute_percentage_error
mean_absolute_percentage_error(y_test, y_pred)

0.06485440905142517

Model is almost 94% accurate

# Logistic Regression

In [ ]:
# Step 1: import lib
import pandas as pd

In [ ]:
# Step 2: import dataset
diabetes = pd.read_csv('https://github.com/YBI-Foundation/Dataset/raw/main/Diabetes.csv')

In [ ]:
diabetes.head()

pregnancies  glucose  diastolic  triceps  insulin   bmi    dpf  age  \
0            6      148         72       35        0  33.6  0.627   50   
1            1       85         66       29        0  26.6  0.351   31   
2            8      183         64        0        0  23.3  0.672   32   
3            1       89         66       23       94  28.1  0.167   21   
4            0      137         40       35      168  43.1  2.288   33   

   diabetes  
0         1  
1         0  
2         1  
3         0  
4         1

In [ ]:
diabetes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   pregnancies  768 non-null    int64  
 1   glucose      768 non-null    int64  
 2   diastolic    768 non-null    int64  
 3   triceps      768 non-null    int64  
 4   insulin      768 non-null    int64  
 5   bmi          768 non-null    float64
 6   dpf          768 non-null    float64
 7   age          768 non-null    int64  
 8   diabetes     768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
diabetes.describe()

pregnancies     glucose   diastolic     triceps     insulin  \
count   768.000000  768.000000  768.000000  768.000000  768.000000   
mean      3.845052  120.894531   69.105469   20.536458   79.799479   
std       3.369578   31.972618   19.355807   15.952218  115.244002   
min       0.000000    0.000000    0.000000    0.000000    0.000000   
25%       1.000000   99.000000   62.000000    0.000000    0.000000   
50%       3.000000  117.000000   72.000000   23.000000   30.500000   
75%       6.000000  140.250000   80.000000   32.000000  127.250000   
max      17.000000  199.000000  122.000000   99.000000  846.000000   

              bmi         dpf         age    diabetes  
count  768.000000  768.000000  768.000000  768.000000  
mean    31.992578    0.471876   33.240885    0.348958  
std      7.884160    0.331329   11.760232    0.476951  
min      0.000000    0.078000   21.000000    0.000000  
25%     27.300000    0.243750   24.000000    0.000000  
50%     32.000000    0.372500   29.000000    0.000000  
75%     36.600000    0.626250   41.000000    1.000000  
max     67.100000    2.420000   81.000000    1.000000

In [ ]:
diabetes.columns

Index(['pregnancies', 'glucose', 'diastolic', 'triceps', 'insulin', 'bmi',
       'dpf', 'age', 'diabetes'],
      dtype='object')

In [ ]:
# Step 3: Define y and x
y = diabetes['diabetes']
x = diabetes.drop(['diabetes'], axis=1)

In [ ]:
y.head(3)

0    1
1    0
2    1
Name: diabetes, dtype: int64

In [ ]:
x.head(3)

pregnancies  glucose  diastolic  triceps  insulin   bmi    dpf  age
0            6      148         72       35        0  33.6  0.627   50
1            1       85         66       29        0  26.6  0.351   31
2            8      183         64        0        0  23.3  0.672   32

In [ ]:
# Step 4: Train test split
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, random_state=2)

In [ ]:
x_train.shape

(576, 8)

In [ ]:
y_train.shape

(576,)

In [ ]:
# Step 5: Select a model
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=200)  # max_iter is 100 by default

# Accuracy: 76.5%

In [ ]:
# Using KNN
# from sklearn.neighbors import KNeighborsClassifier
# model = KNeighborsClassifier()

In [ ]:
# Using Decision Tree
# from sklearn.tree import DecisionTreeClassifier
# model = DecisionTreeClassifier()

In [ ]:
# Using Decision Tree
# from sklearn.ensemble import RandomForestClassifier
# model = RandomForestClassifier()

# Accuracy: 75%

In [ ]:
model.fit(x_train, y_train)

LogisticRegression(max_iter=200)

In [ ]:
# Step 7: Predict
y_pred = model.predict(x_test)

In [ ]:
# Step 8: Accuracy Score
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [ ]:
accuracy_score(y_test, y_pred)

0.765625

In [ ]:
confusion_matrix(y_test,y_pred)

array([[120,  14],
       [ 31,  27]])

In [ ]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.79      0.90      0.84       134
           1       0.66      0.47      0.55        58

    accuracy                           0.77       192
   macro avg       0.73      0.68      0.69       192
weighted avg       0.75      0.77      0.75       192

